runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -0,0 +1,435 @@
+import org.apache.beam.sdk.values.TupleTag;
+import org.joda.time.Instant;
+
+/**
+ * A utility transform that executes a <a
+ * href="https://s.apache.org/splittable-do-fn">splittable</a> {@link DoFn} by expanding it into a
+ * network of simpler transforms:
+ *
+ * <ol>
+ * <li>Pair each element with an initial restriction
+ * <li>Split each restriction into sub-restrictions
+ * <li>Assign a unique key to each element/restriction pair
+ * <li>Group by key (so that work is partitioned by key and we can access state/timers)
+ * <li>Process each keyed element/restriction pair with the splittable {@link DoFn}'s {@link
+ *     DoFn.ProcessElement} method, using state and timers API.
+ * </ol>
+ *
+ * <p>This transform is intended as a helper for internal use by runners when implementing {@code
+ * ParDo.of(splittable DoFn)}, but not for direct use by pipeline writers.
+ */
+@Experimental(Experimental.Kind.SPLITTABLE_DO_FN)
+public class SplittableParDo<
+        InputT, OutputT, RestrictionT, TrackerT extends RestrictionTracker<RestrictionT>>
+    extends PTransform<PCollection<InputT>, PCollection<OutputT>> {
+  private final DoFn<InputT, OutputT> fn;
+  private final DoFnSignature signature;
+
+  /**
+   * Creates the transform for the given original {@link ParDo} and {@link DoFn}.
+   *
+   * @param fn The splittable {@link DoFn} inside the original {@link ParDo} transform.
+   */
+  public SplittableParDo(DoFn<InputT, OutputT> fn) {
+    checkNotNull(fn, "fn must not be null");
+    this.fn = fn;
+    this.signature = DoFnSignatures.INSTANCE.getOrParseSignature(fn.getClass());
+    checkArgument(signature.processElement().isSplittable(), "fn must be a splittable DoFn");
+  }
+
+  @Override
+  public PCollection<OutputT> apply(PCollection<InputT> input) {
+    PCollection.IsBounded isFnBounded = signature.isBounded();
+    Coder<RestrictionT> restrictionCoder =
+        DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn).invokeGetRestrictionCoder();
+    Coder<ElementRestriction<InputT, RestrictionT>> splitCoder =
+        ElementRestrictionCoder.of(input.getCoder(), restrictionCoder);
+
+    PCollection<KeyedWorkItem<String, ElementRestriction<InputT, RestrictionT>>> keyedWorkItems =
+        input
+            .apply(
+                "Pair with initial restriction",
+                ParDo.of(new PairWithRestrictionFn<InputT, OutputT, RestrictionT>(fn)))
+            .setCoder(splitCoder)
+            .apply("Split restriction", ParDo.of(new SplitRestrictionFn<InputT, RestrictionT>(fn)))
+            .setCoder(splitCoder)
+            .apply(
+                "Assign unique key",
+                WithKeys.of(new RandomUniqueKeyFn<ElementRestriction<InputT, RestrictionT>>()))
+            .apply(
+                "Group by key",
+                new GBKIntoKeyedWorkItems<String, ElementRestriction<InputT, RestrictionT>>());
+    checkArgument(
+        keyedWorkItems.getWindowingStrategy().getWindowFn() instanceof GlobalWindows,
```

Comment:
I think we could move this earlier -- we could check the windowing strategy of the input and/or check this in validate (rather than apply).

Everything within here is preserving the original windowing.